src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.CompilationTracker.cs

```diff
@@ -119,78 +120,104 @@ SymbolKind.NetModule or
+                    forkedTrackerState,
                     this.SkeletonReferenceCache.Clone());
 
-                static CompilationTrackerState ForkState(
+                static CompilationTrackerState ForkTrackerState(
                     ProjectState oldProjectState,
                     CompilationTrackerState state,
                     CompilationAndGeneratorDriverTranslationAction? translate)
                 {
-                    if (state is WithCompilationTrackerState withCompilationState)
+                    if (state is WithCompilationTrackerState withCompilationTrackerState)
                     {
-                        var intermediateProjects = state is InProgressState inProgressState
-                            ? inProgressState.IntermediateProjects
-                            : [];
-                        var finalCompilationWithGeneratedDocuments = state is FinalState finalState
-                            ? finalState.FinalCompilationWithGeneratedDocuments
-                            : null;
-
-                        if (translate is not null)
-                        {
-                            // We have a translation action; are we able to merge it with the prior one?
-                            var merged = false;
-                            if (!intermediateProjects.IsEmpty)
-                            {
-                                var (priorState, priorAction) = intermediateProjects.Last();
-                                var mergedTranslation = translate.TryMergeWithPrior(priorAction);
-                                if (mergedTranslation != null)
-                                {
-                                    // We can replace the prior action with this new one
-                                    intermediateProjects = intermediateProjects.SetItem(
-                                        intermediateProjects.Count - 1,
-                                        (oldState: priorState, mergedTranslation));
-                                    merged = true;
-                                }
-                            }
-
-                            if (!merged)
-                            {
-                                // Just add it to the end
-                                intermediateProjects = intermediateProjects.Add((oldProjectState, translate));
-                            }
-                        }
-
-                        var newState = CompilationTrackerState.Create(
-                            withCompilationState.CompilationWithoutGeneratedDocuments,
-                            state.GeneratorInfo,
-                            finalCompilationWithGeneratedDocuments,
-                            intermediateProjects);
-                        return newState;
+                        return ForkWithCompilationTrackerState(oldProjectState, withCompilationTrackerState, translate);
```

Comment:
extracting helpers made the input/output types clearer, and made the logic/flow simpler.